### DISCUSSION TITLE
How to make callback and property bindings work?

### DISCUSSION THREAD

**Discussion by EMUNES** [slint-ui/slint#3662] [Q&A] [answered]
I'm new to Slint and I'm following the rust tutorial and write some code test state bindings:

```
fn main() {
    let app = MainWindow::new().unwrap();
    let app_weak = app.as_weak();

    app.on_add_value(move || {
        eprintln!("--------------------------------");
        let app = app_weak.unwrap();
        app.set_value(2);
    });
    MainWindow::new().unwrap().run().unwrap();
}

slint::slint! {

    import { Button } from "std-widgets.slint";

    export component MainWindow inherits Window {
        width: 326px;
        height: 326px;

        in-out property <int> value;

        callback add_value();

        Text {text: root.value;}

        Button {clicked => {root.add_value()}}

    }
}
```

The app should change the `value` property and the `Text` widget should display differently when I click the button. However, the `Text` always display 0,  so I think it's not working.

I'm trying to use Slint for a more complexing application. For example, in the callback, a rust service will pull video frame data streamed from other devices and the data property will change accordingly. Then the slint UI will show different UI content based on the data property as the streaming continues. But I fail in the simple example above. Anyone has any idea about how to make callback and perperty bindings work properly? Thanks in advance.

I 've checked the coumentation, the tutorial code and searched for solutions. But I just dont know where the problem is. I've used `global callback` to wrap `value` and `add_value()` but it never works, too.

**Comment by tronical** [comment, marked as answer]
Ah that's a small little bug there :-). Change this line:

```rust
MainWindow::new().unwrap().run().unwrap();
```

to read

```rust
app.run().unwrap();
```

and you'll find that it works :-)

The callback that was installed on app was never used because app itself was never used, instead a new instance was created and run.

**Reply by EMUNES** [reply]
Thanks for markting it out. I totally lost focus on this. I'm glad it works well.
